include/ginkgo/core/base/machine_topology.hpp

```diff
@@ -0,0 +1,399 @@
+ *
+ * @note A global object of MachineTopology type is created when first needed
+ * and only destroyed at the end of the program. This means that any subsequent
+ * queries will be from the same global object and hence use an extra atomic
+ * read.
+ */
+class MachineTopology {
+    template <typename T>
+    using hwloc_manager = std::unique_ptr<T, std::function<void(T *)>>;
+
+private:
+    /**
+     * This struct holds the attributes for a normal non-IO object.
+     */
+    struct normal_obj_info {
+        /**
+         * The hwloc object.
+         */
+        hwloc_obj_t obj;
+
+        /**
+         * The logical_id assigned by HWLOC (assigned according to physical
+         * proximity).
+         *
+         * @note Use this rather than os_id for all purposes other than binding.
+         * [Reference](https://www.open-mpi.org/projects/hwloc/doc/v2.4.0/a00364.php#faq_indexes)
+         */
+        size_type logical_id;
+
+        /**
+         * The os_id assigned by the OS (assigned arbitrarily by the OS)
+         */
+        size_type os_id;
+
+        /**
+         * The global persistent id assigned to the object by hwloc.
+         */
+        size_type gp_id;
+
+        /**
+         * The numa number of the object.
+         */
+        int numa;
+
+        /**
+         * The memory size of the object.
+         */
+        size_type memory_size;
+    };
+
+
+    /**
+     * This struct holds the attributes for an IO/Misc object.
+     *
+     * Mainly used for PCI devices. The identifier important for PCI devices is
+     * the PCI Bus ID, stored here as a string. PCI devices themselves usually
+     * contain Hard disks, network components as well as other objects that are
+     * not important for our case.
+     *
+     * In many cases, hwloc is able to identify the OS devices that belong to a
+     * certain PCI Bus ID and here they are stored in the io children vector. A
+     * list of their names are also additionally stored for easy access and
+     * comparison.
+     *
+     * @note IO children can have names such as ibX for Infiniband cards, cudaX
+     * for NVIDIA cards with CUDA and rsmiX for AMD cards.
+     */
+    struct io_obj_info {
+        /**
+         * The hwloc object.
+         */
+        hwloc_obj_t obj;
+
+        /**
+         * The logical_id assigned by HWLOC (assigned according to proximity).
+         *
+         * @note Use this rather than os_id for all purposes other than binding.
+         * [Reference](https://www.open-mpi.org/projects/hwloc/doc/v2.4.0/a00364.php#faq_indexes)
+         */
+        size_type logical_id;
+
+        /**
+         * The os_id assigned by the OS (assigned arbitrarily by the OS)
+         */
+        size_type os_id;
+
+        /**
+         * The global persistent id assigned to the object by hwloc.
+         */
+        size_type gp_id;
+
+        /**
+         * The closest numa.
+         */
+        int numa;
```

Comment:
nit: Why is the closest numa just called `numa` instead of something like `closest_numa`?